Ed25519 keys are not fully supported.

I first noticed this issue trying to self sign a certificate with an Ed25519 key.  Ed25519 keys do not work for self-signing client certificates.

On the client:
```
handshake error: alert: Alert Fatal: BadCertificate
```

On the server:
```
handshake error: dtls fatal: expected and actual key signature do not match
```

But that is just the proximate problem I am having.  Several tests fail if you apply this diff, which switches to ed25519 keys in the `selfsigned` package.

```
diff --git a/pkg/crypto/selfsign/selfsign.go b/pkg/crypto/selfsign/selfsign.go
index 581e0a2..e3b57ac 100644
--- a/pkg/crypto/selfsign/selfsign.go
+++ b/pkg/crypto/selfsign/selfsign.go
@@ -5,7 +5,6 @@ import (
        "crypto"
        "crypto/ecdsa"
        "crypto/ed25519"
-       "crypto/elliptic"
        "crypto/rand"
        "crypto/rsa"
        "crypto/tls"
@@ -20,17 +19,16 @@ var errInvalidPrivateKey = errors.New("selfsign: invalid private key type")

 // GenerateSelfSigned creates a self-signed certificate
 func GenerateSelfSigned() (tls.Certificate, error) {
-       priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+       _, priv, err := ed25519.GenerateKey(rand.Reader)
        if err != nil {
                return tls.Certificate{}, err
        }
-
        return SelfSign(priv)
 }

 // GenerateSelfSignedWithDNS creates a self-signed certificate
 func GenerateSelfSignedWithDNS(cn string, sans ...string) (tls.Certificate, error) {
-       priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+       _, priv, err := ed25519.GenerateKey(rand.Reader)
        if err != nil {
                return tls.Certificate{}, err
        }
```

Failing tests:
```
--- FAIL: TestHandshakeWithAlert (0.41s)
    --- FAIL: TestHandshakeWithAlert/SignatureSchemesNoIntersection (0.00s)
--- FAIL: TestClientCertificate (0.42s)
    --- FAIL: TestClientCertificate/parallel (0.01s)
        --- FAIL: TestClientCertificate/parallel/NoClientCert_cert (0.00s)
        --- FAIL: TestClientCertificate/parallel/RequireAnyClientCert (0.00s)
        --- FAIL: TestClientCertificate/parallel/VerifyClientCertIfGiven_cert (0.00s)
        --- FAIL: TestClientCertificate/parallel/RequireAndVerifyClientCert (0.00s)
        --- FAIL: TestClientCertificate/parallel/RequestClientCert_cert (0.00s)
--- FAIL: TestServerCertificate (0.43s)
    --- FAIL: TestServerCertificate/parallel (0.02s)
        --- FAIL: TestServerCertificate/parallel/good_ca_skip_verify_custom_verify_peer (0.00s)
        --- FAIL: TestServerCertificate/parallel/good_ca_verify_custom_verify_peer (0.00s)
--- FAIL: TestErrorsTemporary (0.00s)
```

--- a/crypto.go
+++ b/crypto.go
@@ -108,6 +108,13 @@ func verifyKeySignature(message, remoteKeySignature []byte, hashAlgorithm hash.A
 // the private key in the certificate.
 // https://tools.ietf.org/html/rfc5246#section-7.3
 func generateCertificateVerify(handshakeBodies []byte, privateKey crypto.PrivateKey, hashAlgorithm hash.Algorithm) ([]byte, error) {
+	if p, ok := privateKey.(ed25519.PrivateKey); ok {
+		// https://pkg.go.dev/crypto/ed25519#PrivateKey.Sign
+		// Sign signs the given message with priv. Ed25519 performs two passes over
+		// messages to be signed and therefore cannot handle pre-hashed messages.
+		return p.Sign(rand.Reader, handshakeBodies, crypto.Hash(0))
+	}
+
 	h := sha256.New()
 	if _, err := h.Write(handshakeBodies); err != nil {
 		return nil, err
@@ -115,9 +122,6 @@ func generateCertificateVerify(handshakeBodies []byte, privateKey crypto.Private
 	hashed := h.Sum(nil)
 
 	switch p := privateKey.(type) {
-	case ed25519.PrivateKey:
-		// https://crypto.stackexchange.com/a/55483
-		return p.Sign(rand.Reader, hashed, crypto.Hash(0))
 	case *ecdsa.PrivateKey:
 		return p.Sign(rand.Reader, hashed, hashAlgorithm.CryptoHash())
 	case *rsa.PrivateKey:

--- a/e2e/e2e_test.go
+++ b/e2e/e2e_test.go
@@ -7,6 +7,7 @@ import (
 	"context"
 	"crypto/ed25519"
 	"crypto/rand"
+	"crypto/rsa"
 	"crypto/tls"
 	"errors"
 	"fmt"
@@ -362,6 +363,127 @@ func testPionE2ESimpleED25519(t *testing.T, server, client func(*comm)) {
 	}
 }
 
+func testPionE2ESimpleED25519ClientCert(t *testing.T, server, client func(*comm)) {
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
+
+	report := test.CheckRoutines(t)
+	defer report()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	_, skey, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		t.Fatal(err)
+	}
+	scert, err := selfsign.SelfSign(skey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, ckey, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ccert, err := selfsign.SelfSign(ckey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	scfg := &dtls.Config{
+		Certificates: []tls.Certificate{scert},
+		CipherSuites: []dtls.CipherSuiteID{dtls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256},
+		ClientAuth:   dtls.RequireAnyClientCert,
+	}
+	ccfg := &dtls.Config{
+		Certificates:       []tls.Certificate{ccert},
+		CipherSuites:       []dtls.CipherSuiteID{dtls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256},
+		InsecureSkipVerify: true,
+	}
+	serverPort := randomPort(t)
+	comm := newComm(ctx, ccfg, scfg, serverPort, server, client)
+	comm.assert(t)
+}
+
+func testPionE2ESimpleECDSAClientCert(t *testing.T, server, client func(*comm)) {
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
+
+	report := test.CheckRoutines(t)
+	defer report()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	scert, err := selfsign.GenerateSelfSigned()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ccert, err := selfsign.GenerateSelfSigned()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	scfg := &dtls.Config{
+		Certificates: []tls.Certificate{scert},
+		CipherSuites: []dtls.CipherSuiteID{dtls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256},
+		ClientAuth:   dtls.RequireAnyClientCert,
+	}
+	ccfg := &dtls.Config{
+		Certificates:       []tls.Certificate{ccert},
+		CipherSuites:       []dtls.CipherSuiteID{dtls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256},
+		InsecureSkipVerify: true,
+	}
+	serverPort := randomPort(t)
+	comm := newComm(ctx, ccfg, scfg, serverPort, server, client)
+	comm.assert(t)
+}
+
+func testPionE2ESimpleRSAClientCert(t *testing.T, server, client func(*comm)) {
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
+
+	report := test.CheckRoutines(t)
+	defer report()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	spriv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatal(err)
+	}
+	scert, err := selfsign.SelfSign(spriv)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cpriv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ccert, err := selfsign.SelfSign(cpriv)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	scfg := &dtls.Config{
+		Certificates: []tls.Certificate{scert},
+		CipherSuites: []dtls.CipherSuiteID{dtls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256},
+		ClientAuth:   dtls.RequireAnyClientCert,
+	}
+	ccfg := &dtls.Config{
+		Certificates:       []tls.Certificate{ccert},
+		CipherSuites:       []dtls.CipherSuiteID{dtls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256},
+		InsecureSkipVerify: true,
+	}
+	serverPort := randomPort(t)
+	comm := newComm(ctx, ccfg, scfg, serverPort, server, client)
+	comm.assert(t)
+}
+
 func TestPionE2ESimple(t *testing.T) {
 	testPionE2ESimple(t, serverPion, clientPion)
 }
@@ -376,4 +498,16 @@ func TestPionE2EMTUs(t *testing.T) {
 
 func TestPionE2ESimpleED25519(t *testing.T) {
 	testPionE2ESimpleED25519(t, serverPion, clientPion)
+}
+
+func TestPionE2ESimpleED25519ClientCert(t *testing.T) {
+	testPionE2ESimpleED25519ClientCert(t, serverPion, clientPion)
+}
+
+func TestPionE2ESimpleECDSAClientCert(t *testing.T) {
+	testPionE2ESimpleECDSAClientCert(t, serverPion, clientPion)
+}
+
+func TestPionE2ESimpleRSAClientCert(t *testing.T) {
+	testPionE2ESimpleRSAClientCert(t, serverPion, clientPion)
 }

--- a/pkg/crypto/selfsign/selfsign.go
+++ b/pkg/crypto/selfsign/selfsign.go
@@ -72,6 +72,11 @@ func WithDNS(key crypto.PrivateKey, cn string, sans ...string) (tls.Certificate,
 	names := []string{cn}
 	names = append(names, sans...)
 
+	keyUsage := x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign
+	if _, isRSA := key.(*rsa.PrivateKey); isRSA {
+		keyUsage |= x509.KeyUsageKeyEncipherment
+	}
+
 	template := x509.Certificate{
 		ExtKeyUsage: []x509.ExtKeyUsage{
 			x509.ExtKeyUsageClientAuth,
@@ -79,7 +84,7 @@ func WithDNS(key crypto.PrivateKey, cn string, sans ...string) (tls.Certificate,
 		},
 		BasicConstraintsValid: true,
 		NotBefore:             time.Now(),
-		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		KeyUsage:              keyUsage,
 		NotAfter:              time.Now().AddDate(0, 1, 0),
 		SerialNumber:          serialNumber,
 		Version:               2,

